### ISSUE TITLE
OOM 2nd iteration of OPT-66B inference

### ISSUE THREAD

**Issue by andre-bauer** [huggingface/accelerate#1882]
### System Info

```Shell
- `Accelerate` version: 0.21.0
- Platform: Linux-5.15.0-1013-gcp-x86_64-with-glibc2.31
- Python version: 3.9.16
- Numpy version: 1.23.5
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 204.45 GB
- GPU type: Tesla T4
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Script I use:
```python
import gc
import re

import click
import torch
from accelerate import init_empty_weights, infer_auto_device_map, load_checkpoint_and_dispatch
from datasets import load_dataset
from huggingface_hub import snapshot_download
from tqdm import tqdm
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig

@click.command()
@click.option("--model-type", default="facebook/opt-66b", type=str)
@click.option("--cache-dir", default=".cache/hf_cache", type=str)
@click.option("--dataset", default="wikitext", type=str)
@click.option("--eval_fraction", default=1.0, type=float)
def main(model_type, cache_dir, dataset, eval_fraction):
    with open('/proc/meminfo') as f:
        meminfo = f.read()
    matched = re.search(r'^MemTotal:\s+(\d+)', meminfo)
    if matched:
        mem_total_kB = int(matched.groups()[0])

    content_root = snapshot_download(
        model_type,
        ignore_patterns=[
            "*.msgpack",
            "*.h5",
            "*.msgpack.index.json",
            "*.h5.index.json",
            "*.safetensors",
            "*.safetensors.index.json",
        ],
        cache_dir=cache_dir,
    )

    config = AutoConfig.from_pretrained(model_type)
    with init_empty_weights():
        model = AutoModelForCausalLM.from_config(config, torch_dtype=torch.float16)
    model.tie_weights()

    # Hack because some opt version have some quirks, not sure why
    if model_type in ["facebook/opt-6.7"]:
        dist_model = model.model
    else:
        dist_model = model

    max_memory = {
        **{i: torch.cuda.get_device_properties(i).total_memory * (0.4 if i == 0 else 0.8) for i in range(torch.cuda.device_count())},
        "cpu": f"{mem_total_kB}kB",
    }
    print(max_memory)
    device_map = infer_auto_device_map(
        dist_model, max_memory=max_memory, no_split_module_classes=["OPTDecoderLayer"], dtype=torch.float16, verbose=True
    )
    print(device_map)
    load_checkpoint_and_dispatch(
        dist_model, checkpoint=content_root, device_map=device_map
    )
    model.tie_weights()
    tokenizer = AutoTokenizer.from_pretrained(model_type, use_fast=False)
    print(model)

    if dataset == "wikitext":
        model.eval()

        test = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
        encodings = tokenizer("\n\n".join(test["text"]), return_tensors="pt")

        max_length = model.config.max_position_embeddings
        stride = 1024
        seq_len = encodings.input_ids.size(1)
        encodings.input_ids = encodings.input_ids[:, : int(seq_len * eval_fraction)]
        nlls = []
        prev_end_loc = 0
        for begin_loc in tqdm(range(0, seq_len, stride)):
            end_loc = min(begin_loc + max_length, seq_len)
            trg_len = end_loc - prev_end_loc  # may be different from stride on last loop
            input_ids = encodings.input_ids[:, begin_loc:end_loc].to("cuda")
            target_ids = input_ids.clone()
            target_ids[:, :-trg_len] = -100

            torch.cuda.empty_cache()
            with torch.no_grad():
                outputs = model(input_ids, labels=target_ids)
                neg_log_likelihood = outputs.loss
            gc.collect()
            torch.cuda.empty_cache()
            nlls.append(neg_log_likelihood)
            prev_end_loc = end_loc
            if end_loc == seq_len:
                break

        ppl = torch.exp(torch.stack(nlls).mean())
        print(ppl)
```

### Expected behavior

Iteration 1 works properly, but the 2nd iteration gives me an OOM at softmax. I wonder if something does not get moved back to cpu correctly?

**Comment by andre-bauer** [comment]
If i manually set everything in the device map to cpu except `device_map["model.decoder.final_layer_norm"] = 0` it works 🤔 Which feels wrong, before that `layers.0` was also on device 0

**Comment by SunMarc** [comment]
Hi @andre-bauer , can I have the full traceback ? Also, you can and should directly initialize your model with `AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16, device_map="auto")` or if you want to pass max_memory: `AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16, max_memory=custom_max_memory)`.
